First commit 「Sinatra を使ってWebアプリケーションの基本を理解する」

yumiya-g opened the pull request.

yumiya-g committed:
First commit 「Sinatra を使ってWebアプリケーションの基本を理解する」

JunichiIto commented on README.md:
```diff
@@ -1,2 +1,68 @@
+
+2. Gemfileを確認
+
+Gemfileが下記のとおりになっていることを確認します。
```
GitHubで管理されているし、Gemfileを更新したらREADMEも更新するの？っていう話になりそうなのでわざわざ書かなくていいかな〜と思います。

yumiya-g replied:
不要な記述ですので、削除します。

JunichiIto commented on README.md:
````diff
@@ -1,2 +1,68 @@
+bundle install
+```
+
+実行後、`Gemfile.lock`が生成されたことを確認します。
````
これもGitHub管理なので確認不要です〜。

yumiya-g replied:
こちらも不要な記述ですので、削除します。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,162 @@
+
+# 提出時に削除する
+require 'sinatra/reloader'
+require 'debug'
```
削除し忘れ？

yumiya-g replied:
削除漏れです。失礼いたしました。。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,162 @@
+    @memo_ids = @memo_lists.map(&:keys).flatten
+  end
+
+  def generate_memo_lists
```
generate=生成するというメソッドではないなと思いました。
もっといい名前を検討してみてください〜。

yumiya-g replied:
生成された「メモ情報を表示」するメソッド名に変更します。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,162 @@
+  end
+
+  def create_new_memo(title, content)
+    # latest_idキーからvalueを取得して、カウントをプラス１（初回投稿時には、ID1を作成）する
```
コメントが冗長なので不要なものは削除してください。
https://qiita.com/jnchito/items/f0d90af4ed44b7484103

yumiya-g replied:
こちらも削除漏れでした。。失礼いたしました。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,162 @@
+
+  def create_new_memo(title, content)
+    # latest_idキーからvalueを取得して、カウントをプラス１（初回投稿時には、ID1を作成）する
+    @latest_memo_id = @latest_memo_id.zero? ? 1 : @latest_memo_id + 1
```
0 + 1 = 1なので条件分岐は不要そうです。

```suggestion
    @latest_memo_id += 1
```

yumiya-g replied:
確かに初期値が「0」なので、条件分岐は不要ですね。
修正いたします。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,162 @@
+    @memos_data = @memo_lists.push new_memo
+
+    # データを追加したhashを、JSONファイル形式に変換
+    updated_memos = JSON.generate({ "latest_id": @latest_memo_id, "memo_lists": @memos_data })
```
`"latest_id": @latest_memo_id`とJSONっぽいフォーマットで書いていますが、この時点ではRubyのハッシュなので、`latest_id: @latest_memo_id`とやっていることは同じです。

yumiya-g replied:
ハッシュのキーにダブルクォーテーションは不要ですね。
こちらも削除いたします。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,162 @@
+      { title: @title, content: @content, existing_id: @existing_id }
+    end
+
+    memo.compact.first
```
mapしてcompactしてfirst、はfindメソッドに置き換えることが可能です。

yumiya-g replied:
findメソッドで処理できるよう修正します。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,162 @@
+  end
+
+  def show_memo_details(memo_id)
+    return unless @memo_ids.include?(memo_id)
```
この行はあってもなくても同じ結果になると思うのですがいかがでしょう？
（updateやdeleteも同様）

yumiya-g replied:
こちらは、該当するメモIDがデータベースに含まれているかを事前にチェックするため、導入していました。
仰るとおりなくても差し支えないので削除いたします。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,162 @@
+end
+
+get '/memos' do
+  @memos = Memo.parse(DB_FILE_NAME).generate_memo_lists
```
どのメソッド呼び出しも

```ruby
Memo.parse(DB_FILE_NAME).xxx
```

になっていますが、`.parse(DB_FILE_NAME)`が繰り返しになっていて冗長なので、

```ruby
Memo.generate_memo_lists
```

と呼び出せるようにした方がいいと思います。

あと、Memoクラスは「1インスタンス = 1件のメモ」ではなく、Memoのデータベースとしての役割になってそうなので、

```ruby
MemoDB.xxx
```

みたいな名前が良いと思います。

yumiya-g replied:
クラス名の見直し、およびデータベースの呼び出し方法を修正します。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,162 @@
+end
+
+get '/memos/:id' do
+  @memos = Memo.parse(DB_FILE_NAME).show_memo_details(@params['id'])
```
複数形だと配列と勘違いされそうです。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,162 @@
+    memo = @memo_lists.map do |m|
+      next unless m.keys.first == memo_id
+
+      @title = m.values.first['title']
```
インスタンス変数でなく、ローカル変数でよいのでは？
インスタンス変数は寿命が長く、どこで読み書きされるのかわからないため、多用するとカオスになりがちです。
インスタンス変数は必要最小限の使用に抑えてください。

yumiya-g replied:
> インスタンス変数は寿命が長く、どこで読み書きされるのかわからないため、多用するとカオスになりがちです。

ご指摘どおりローカル変数で問題ないので修正を行います。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,162 @@
+      file.write(updated_memos)
+    end
+  end
+
+  def self.parse(filename)
+    file = File.open(filename, 'r')
+    memo_data = JSON.parse(file.read)
+    Memo.new(memo_data)
+  end
```
```suggestion
  private
  
  def self.parse
    file = File.open(DB_FILE_NAME, 'r')
    JSON.parse(file.read)
  end
```

のような実装にして、各メソッドはその都度parseメソッドを呼び出すようにしましょう。
別のコメントでも書きましたが、インスタンス変数は使いすぎに要注意です。
Sinatraであればあくまでerbとデータをやりとりするときだけ使う、というふうにしましょう。

yumiya-g replied:
> 各メソッドはその都度parseメソッドを呼び出すようにしましょう。

`private`は不要とのことでしたので、クラスメソッドの変更と各メソッドの調整を行います。

JunichiIto reviewed (commented):
コメントしました〜。

yumiya-g committed:
Fix: 不要な作業指示を削除し、説明文を修正しました

yumiya-g committed:
Fix: メソッド名の見直しと、show_memo_listsメソッドの処理を修正しました

- メモを一覧表示するメソッド名を「show_memo_lists」に変更
- 最新IDを生成する際の条件分岐は不要のため削除
- インスタンス変数「@memos_data」をローカル変数に変更
- JSON.generateの引数となるハッシュのキーから、ダブルクォーテーションを削除

yumiya-g committed:
Fix: show_memo_detailsメソッドを修正しました

- 「@memo_ids」を削除
  - 表示するメモIDとデータベースに保存されているメモIDの照合が不要のため
  - その他のメソッド内の照合処理も削除
- メモの詳細データを取得する処理をfindメソッドに変更し、メモIDから直接メモのタイトルと内容を取得する

yumiya-g committed:
Fix: Memoクラスのクラス名変更と、クラスメソッドの処理を修正しました

- クラス名を「MemoDB」に変更
- クラスメソッド名を「parse」から「generate_memo_lists」に変更
  - クラスメソッドから引数を除外し、定数からファイルを呼び出す
- showテンプレートとeditテンプレートに渡すインスタンス変数名を単数系に変更
  - @memos -> @memo
- patchとdeleteで、インスタンス変数に代入を行わない処理に変更
  - DB書き換え処理を行うため、代入は不要

yumiya-g committed:
Fix: インスタンス変数名を単数形に変更しました

- main.rbで変数名を変更したため、合わせて修正
- main.rb 修正箇所
  - ブロックパラメータ名の変更
  - ハッシュのキーからダブルクォーテーションを削除

yumiya-g committed:
Add: メモ一覧ページに条件分岐を加えました

- メモデータが無い（db.json："memo_lists"の値の配列が空）の場合、メッセージを表示する

yumiya-g committed:
Fix: クラス内のインスタンスメソッドを、クラスメソッドに変更しました

- メソッドの呼び出しはクラスメソッドのみとし、各クラスメソッド内でJSONのパース（generate_memo_lists）を実行
- 変数名等の微調整
- 開発用のGemを削除

yumiya-g commented:
@JunichiIto さん

レビューへの対応を行いました。
再度ご確認のほどをお願いします。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,133 @@
+require 'json'
+require 'sinatra'
+
+DB_FILE_NAME = 'db.json'
```
MemoDBの中でしか使わないので、クラス定義内に移動しましょう

yumiya-g replied:
`MemoDB`クラス内に記載します。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,133 @@
+end
+
+class MemoDB
+  def self.show_memo_lists
```
このメソッド自身が何かを表示するわけではないので、showは不自然な気がします。
データを読み取る＝readとかの方が良いと思います。
あと、memo lists = memos と表現しても問題なさそうです。
なので、直すとすれば`read_memos`とかでしょうか。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,133 @@
+    generate_memo_lists['memo_lists']
+  end
+
+  def self.create_new_memo(title, content)
```
create_old_memoがあるとは思えないので、create_memoだけでも良さそうです。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,133 @@
+
+  def self.create_new_memo(title, content)
+    latest_id = generate_memo_lists['latest_id'] += 1
+    memo_lists = generate_memo_lists['memo_lists']
```
`generate_memo_lists`でファイルを2回読み取るのは非効率です。
ほとんど体感できないかもしれませんが、ファイルの読み書きは比較的コストの高い処理になります。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,133 @@
+    memo_lists = generate_memo_lists['memo_lists']
+
+    new_memo = { latest_id => { 'title' => title, 'content' => content } }
+    updated_lists = memo_lists.push new_memo
```
pushは破壊的メソッドなので、新しい変数に代入してもあまり意味がないですね。
どうしても非破壊的にしたいなら、

```ruby
updated_lists = [*memo_lists, new_memo]
```

とかですが、ここは単純に

```suggestion
    memo_lists.push new_memo
```

だけでいいかなと思います。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,133 @@
+    save_memo_data(deleted_memos)
+  end
+
+  def self.save_memo_data(updated_memos)
```
このメソッドはハッシュを受け取って、この中でハッシュをJSONに変換した方がDRYになりそうです。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,133 @@
+    save_memo_data(updated_memos)
+  end
+
+  def self.show_memo_details(memo_id)
```
read_memosは複数のメモを返す、とするなら、こっちはread_memoで1件のメモを返す、という命名が使えそうです

JunichiIto commented on db.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "latest_id": 0,
+  "memo_lists": []
```
```suggestion
  "memo_lists": {}
```

として、以下のような形式（注：JSONでなくRubyで書いてます）でデータを管理した方がいろいろ効率が良さそうです。

```ruby
memo_lists: {
  1 => {title: 'hello', content: 'hi'},
  2 => {title: 'farewell', content: 'bye'},
}
```

JunichiIto commented on views/header.erb:
```diff
@@ -0,0 +1,6 @@
+  <header>
+    <h1><a href="/"><%= @app_title %></a></h1>
+    <% if @env["REQUEST_PATH"] == "/memos" %>
```
views/index.erb にリンクを載せるようにすればこの条件分岐は不要になりそうだなと思いました。

JunichiIto commented on views/show.erb:
```diff
@@ -0,0 +1,22 @@
+
+      <div class="col">
+        <ul class="container">
+          <li><button onclick="location.href='<%= @params["id"] %>/edit'">変更</button></li>
```
buttonよりもaタグで実装した方が自然ですね。buttonだとわざわざJSを書かないといけないので。

JunichiIto reviewed (commented):
コメントしました〜。

yumiya-g committed:
Fix: 各メソッド内の変数名を調整し、JSON生成処理をDRYにしました

- DB名を格納した定数をクラス内に定義
- generate_memo_listsメソッドの結果を変数memosに格納してから、メモIDとデータを取得する
- 各メソッドで生成されたハッシュを、save_memo_dataメソッドに渡してJSON形式に変換する
- 各メソッド名、および各メソッド内の変数名を見直し

yumiya-g committed:
Change: memo_listsキーに対応する値を配列からオブジェクトに変更しました

- 配列では全件検索が行われるので、オブジェクトに変更してメモIDからデータ取得を行うため

yumiya-g committed:
Fix: メモデータベースの構造変更に伴う、追加・詳細表示・更新・削除処理を修正しました

- db.jsonの「memo_lists」変数の値をオブジェクトに変更したため

yumiya-g committed:
Fix: ヘッダーエリア「追加する」ボタンを調整しました

- index.erb
  - 条件分岐を削除して、TOPページ（メモ一覧）のみボタンを表示
  - TOPページにはフォームが無いため、buttonタグからaタグに変更
- show.erb
  -「変更」ボタンをbuttonタグからaタグに変更

yumiya-g committed:
Delete: 開発用Gemを削除

yumiya-g commented:
@JunichiIto さん

レビューを頂き、再度修正を行いました。
ご確認くださいますようお願いします。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,130 @@
+    memo = memos[memo_id]
+    return nil if memo.nil?
+
+    { title: memo['title'], content: memo['content'] }
```
```suggestion
    memos[memo_id]
```

ハッシュから別のハッシュを作るより、取得したハッシュをそのまま返せばいいのでは？と思いました。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,130 @@
+    latest_id = memos['latest_id']
+    memo_lists = memos['memo_lists']
+
+    memo_lists[params['id']] = { 'title' => params['title'], 'content' => params['content'] }
```
sliceメソッドを使うといいかも？

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,130 @@
+    new_memo = { 'title' => title, 'content' => content }
+    memos['memo_lists'][latest_id] = new_memo
+    changed_memos = { latest_id:, memo_lists: }
+    save_memo_data(changed_memos)
```
```suggestion
    save_memo_data(latest_id:, memo_lists:)
```

だけでもいいかも。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,130 @@
+    converted_memos = JSON.generate(changed_memos)
+
+    File.open(DB_FILE_NAME, 'w') do |file|
+      file.write(converted_memos)
```
```suggestion
    File.open(DB_FILE_NAME, 'w') do |file|
      converted_memos = JSON.generate(changed_memos)
      file.write(converted_memos)
```

とした方が変数の寿命が短くなりそうです。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,130 @@
+
+  def self.generate_memo_lists
+    file = File.open(DB_FILE_NAME, 'r')
+    JSON.parse(file.read)
```
参考：キーをシンボルにするオプションもあります。
https://docs.ruby-lang.org/ja/latest/method/JSON/m/parse.html

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,130 @@
+    end
+  end
+
+  def self.generate_memo_lists
```
generateは生成するの意味です。「ファイルからデータを読み取ってハッシュを生成する」と解釈できなくもないですが、それは内部的な処理であって、メソッドの利用者から見れば「データを取ってくる」とか「データを読み取る」といった処理に見えるはずです。そういうニュアンスのメソッド名に変えてみてください。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,130 @@
+  def self.delete_memo(params)
+    memos = generate_memo_lists
+    latest_id = memos['latest_id']
+    memo_lists = memos['memo_lists']
```
generate_memo_listsを代入するのはmemosで、memos['memo_lists']を代入するのがmemo_listsなんですね。
どのデータを何と呼ぶのか、一貫性を持たせた方がいいです。

[ユーザーを見つけたら……Aさん「自転車に入れます」Bさん「人間に入れます」Cさん「ユーザーに入れます」 \| FBC](https://bootcamp.fjord.jp/pages/readable-variable-name)

ちなみに`_list`は配列の意味を持つことが多いので、`memo_lists`がハッシュなのは誤解を招きそうです。
一方、`memos`のようにただの配列の場合は、「配列またはハッシュ」として解釈されることが多いです。
あと、`memo_list`と`memos`は同じ意味になるので、後者の名前を使う方がいいです。

・・・と、コメントすると「いったい、どう命名すれば！？」と頭を抱えてしまうかもしれませんが、じっくり考えてみてくださいw

JunichiIto commented on views/index.erb:
```diff
@@ -0,0 +1,17 @@
+    <p>メモはありません</p>
+    <% else %>
+    <ul>
+      <% @memos.map do |memo| %>
```
```suggestion
      <% @memos.map do |x, y| %>
```

のようにキーと値を別々の変数で受ける方がコードがシンプルになります。（パラメータ名は自分で考えてください）

JunichiIto commented on views/show.erb:
```diff
@@ -0,0 +1,22 @@
+          <li>
+            <form method="post" action="/memos/<%= @params["id"] %>/delete">
+              <input type="hidden" name="_method" value="delete">
+              <button onclick="location.href='<%= @params["id"] %>/delete'">削除</button>
```
onclick属性は不要なのでは？

JunichiIto commented on views/edit.erb:
```diff
@@ -0,0 +1,22 @@
+      <p>ID:<%= @params["id"] %></p>
+      <div class="col">
+        <form method="post" action="/memos/<%= @params["id"] %>/edit">
+        <input type="hidden" name="_method" value="patch">
```
```suggestion
          <input type="hidden" name="_method" value="patch">
```

かな？

JunichiIto reviewed (commented):
コメントしました〜。

yumiya-g committed:
Fix: 各テンプレートファイルを修正

- edit：インデント調整
- index：ブロックパラメータにハッシュのキーと値を指定して、データの取得処理を修正
- show：button要素からonclick属性を削除

yumiya-g committed:
Fix: クラスメソッドread_memoを修正

- ハッシュの再作成を見直す
- 不要な条件分岐を削除

yumiya-g committed:
Fix: メモ作成時とJSONへの保存処理を修正

create_memoメソッド：新規追加後のデータを変数に代入せず、save_memo_dataメソッドに直接渡す
save_memo_dataメソッド：仮引数を2つにし、ファイルへの書き込み直前にJSON生成処理を行う

yumiya-g committed:
Fix: メモの更新と削除処理を見直しました

- update_memo_details：sliceメソッドで実装
- delete_memo
  - save_memo_dataメソッドに未削除のメモIDと本文を渡す
  - sliceメソッドで、paramsから受け取ったタイトルとメモ内容を抽出

yumiya-g committed:
Fix: generate_memo_listsメソッドを修正しました

- メソッド名をload_memosへ変更
- parseメソッドのsymbolize_namesオプションで、生成するハッシュのキーを文字列からシンボルに変換

yumiya-g committed:
Fix: JSONデータのキーをシンボルに変更後、表示の調整を行いました

- ハッシュから値を取得する処理を文字列からシンボルへ変更
- メモを保存するメソッド名をsave_memo_data -> save_memo に変更

yumiya-g committed:
Fix: メモタイトルと本文を格納する値に紐づく、キー名を変更しました

yumiya-g committed:
Fix: JSONのキー名を変更後、対応する値の取得処理と変数名の変更を行いました

yumiya-g committed:
Remove: JSON内の初期値を削除しました

- 登録されるメモデータ構造を修正するため

yumiya-g committed:
Improve: メモIDをUUIDに変更して、処理変更を行いました

- データベース構造は下記に変更
  - UUID=>{:title=> "タイトル", content=> "本文"}

yumiya-g committed:
Fix: デバッグ用ツール削除、メモ一覧ページテンプレートを修正しました

- メモ一覧ページテンプレート（index.erb）の、ブロックパラメータ名をdb.jsonと一致するよう修正

yumiya-g commented:
@JunichiIto さん

レビューいただきありがとうございました。
再提出いたしましたので、ご確認のほどをお願いします。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,117 @@
+  DB_FILE_NAME = 'db.json'
+
+  def self.read_memos
+    load_memos
```
read_memosがload_memosのエイリアスメソッドになってるだけなので、直接load_memosを呼べばいいと思いますー。

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,117 @@
+    memos = load_memos
+    id = SecureRandom.uuid
+
+    memos[id] = { 'title' => title, 'content' => content }
```
読み込み時にキーをシンボルにするなら、ここもシンボルにした方が一貫性があると思います。
（もしくは読み込み時にキーを文字列のままにするのも一手）

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,117 @@
+end
+
+post '/memos/new' do
+  MemoDB.create_memo(@params['title'], @params['content'])
```
updateと同様にハッシュを渡してしまった方がいいかも。
将来的に項目が増えると、引数もどんどん増えていってしまうので。

```suggestion
  MemoDB.create_memo(@params)
```

JunichiIto commented on main.rb:
```diff
@@ -0,0 +1,117 @@
+    memos[id.to_sym]
+  end
+
+  def self.update_memo_details(params)
```
ここだけ_detailsを付ける意味ってあるのかなー？と思いました。

JunichiIto commented on public/style.css:
```diff
@@ -0,0 +1,142 @@
+        }
+      }
+    }
+    & div#form,
```
```suggestion
    & #form,
```

idなのでdivは不要ですね。

それとは別に、styleを当てる場合はidよりもclassを使った方がいいかもしれません。

JunichiIto commented on views/index.erb:
```diff
@@ -0,0 +1,17 @@
+    <% else %>
+    <ul>
+      <% @memos.map do |id, content| %>
+      <li><a href="memos/<%= id.to_s %>"><%= h(content[:title]) %></a></li>
```
```suggestion
      <li><a href="memos/<%= id %>"><%= h(content[:title]) %></a></li>
```

to_sしなくても自動的に文字列変換されると思います。

JunichiIto reviewed (commented):
コメントしました〜。

yumiya-g committed:
Fix: フォームを表示するテンプレートのdivタグをid属性からclass属性に変更しました

- style.css：divを削除して、formクラスに修正
- erbファイル：id属性からclass属性に変更

yumiya-g committed:
Fix: mapメソッド内のキー「id」から、to_sメソッドを除外しました

yumiya-g committed:
Fix: メソッド名の見直し、引数の指定方法、ハッシュのキー形式を修正しました

- エイリアスメソッド`read_memos`を削除
- ハッシュのキーを文字列からシンボルへ変更
- メソッド名`update_memo_details` → `update_memo`へ変更
- `create_memo`メソッドの実引数にハッシュを渡す

JunichiIto reviewed (approved):
確認しました！これでOKです〜🙆‍♂️

yumiya-g commented:
@JunichiIto さん

レビューいただきありがとうございました❗️
